### ISSUE TITLE
Allen&Heath Xone K2 Mapping problem

### ISSUE THREAD

**Issue by uriwilbury** [mixxxdj/mixxx#13144]
### Bug Description

My custom mapping for the controller Allen&Heath Xone K2, which worked until version 2.3.6, is not working for version 2.4.0 and newer.

I get the following error:

“The mapping for your controller “XONE:K2” is not working properly.
The script code needs to be fixed.
The functionality provided by this controller mapping will be disabled until the issue is resolved.”

Details: 

Uncaught exception: file:///home/oriol/.mixxx/controllers/midi-components-0.0.js:139: TypeError: Cannot call method 'trigger' of undefined

Backtrace: @file:///home/oriol/.mixxx/controllers/midi-components-0.0.js:139

trigger@file:///home/oriol/.mixxx/controllers/midi-components-0.0.js:138

@file:///home/oriol/.mixxx/controllers/midi-components-0.0.js:485

applyOperationTo@file:///home/oriol/.mixxx/controllers/midi-components-0.0.js:462

forEachComponent@file:///home/oriol/.mixxx/controllers/midi-components-0.0.js:474

reconnectComponents@file:///home/oriol/.mixxx/controllers/midi-components-0.0.js:479

@file:///home/oriol/.mixxx/controllers/midi-components-0.0.js:1016

@file:///home/oriol/.mixxx/controllers/Allen-and-Heath-Xone-K2-scripts.js:851

@file:///home/oriol/.mixxx/controllers/Allen-and-Heath-Xone-K2-scripts.js:855

@file:///home/oriol/.mixxx/controllers/Allen-and-Heath-Xone-K2-scripts.js:47

### Version

2.4.0, 2.4.1, 2.5.0-alpha

### OS

Linux Mint 21.1 Cinnamon

**Comment by ronso0** [comment]
In order to debug this we need the script file ; )
Please create a gist (or fork Mixxx and create a branch) and share a link here.

**Comment by uriwilbury** [comment]
Not sure if I did it right but there zou go :)
https://gist.github.com/uriwilbury/4c66d2b5419b3c95030f38419bc99e77

**Comment by JoergAtGithub** [comment]
I guess it's the same as #13180 about the official K1/K2 mapping that comes with Mixxx.

**Comment by uriwilbury** [comment]
So it is a known bug with no solution yet, right?

Is there a way to go back to version 2.3.6 until it is fixed?

thanks

**Comment by Swiftb0y** [comment]
Can you try deleting the `~/.mixxx/controllers/midi-components-0.0.js` and try again?

**Comment by Swiftb0y** [comment]
> Is there a way to go back to version 2.3.6 until it is fixed?

You can, it depends on your installation method, but it should be possible to downgrade packages using `apt` (your package manager on Mint I assume). I wouldn't recommend it though. It would get fixed much faster if you stay on 2.4 and properly help us to diagnose the issue.

**Comment by Swiftb0y** [comment]
> I guess it's the same as https://github.com/mixxxdj/mixxx/issues/13180 about the official K1/K2 mapping that comes with Mixxx.

I doubt it. An exception due to accessing `undefined` is much less severe and easier to debug than a crash in QV4's memory code... The former is very likely not a related to the latter.

**Comment by pierrelegall** [comment]
> Uncaught exception: file:///home/oriol/.mixxx/controllers/midi-components-0.0.js:139: TypeError: Cannot call method 'trigger' of undefined

The problem comes from `midi-components-0.0.js`. I suppose you copied it from Mixxx sources (and maybe changed it) as I did.

I've got the same problem upgrading to `v2.4.0` with my K2. Just get the new version of it (and maybe recustomize it). You could find it [there](https://github.com/mixxxdj/mixxx/blob/e32b1ede66169982a421ecdb2135773790e991e4/res/controllers/midi-components-0.0.js).

**Comment by Swiftb0y** [comment]
yes, alternatively if you delete the midi-components from your `~./mixxx/` directory, mixxx will copy the appropriate version into that folder the next time you start it.

**Comment by uriwilbury** [comment]
I deleted the midi-components-0.0.js and now the controller is working!

I still got the following error after starting Mixxx with the controller and pushing a couple buttons:

Uncaught exception: file:///home/oriol/.mixxx/controllers/Allen-and-Heath-Xone-K2-scripts.js:121: TypeError: Cannot read property 'columns' of undefined
Backtrace: @file:///home/oriol/.mixxx/controllers/Allen-and-Heath-Xone-K2-scripts.js:121

This can be ignored though on the message error and the controller seems to still work.

I'll keep pushing buttons to see if there is any other issues.

thanks!

**Comment by Swiftb0y** [comment]
interesting, is that still the same script version as from the gist you posted? Because that code should only run on mapping shutdown.

**Comment by macrophone** [comment]
Same original issue here, I've renamed midi-components-0.0.js in .mixxx/ and nos it seems working fine. 
(note that there is no new midi-components-0.0.js in the .mixxx/ folder after reboot.)

**Comment by Holzhaus** [comment]
Yes, mixxx used to always copy mapping files to `.mixxx` unconditionally upon selecting it. We fixed that (in 2.3 IIRC), exactly because it does not make sense and leads to mappings not being updated and potentially locking people in with ancient mappings. I suppose the copy you had in that directly was a leftover from an older version before the fix.

**Comment by pierrelegall** [comment]
> Yes, mixxx used to always copy mapping files to `.mixxx` unconditionally upon selecting it. We fixed that (in 2.3 IIRC), exactly because it does not make sense and leads to mappings not being updated and potentially locking people in with ancient mappings. I suppose the copy you had in that directly was a leftover from an older version before the fix.

Didn't know it was the previous behaviour. Much better like this :+1:

**Comment by Swiftb0y** [comment]
well, the better issue text would be "controller script libraries not upgraded in `~/.mixxx/` after mixxx upgrade`. This is indeed a general issue we need to fix. We need a mechanism to either version js libraries or to distinguish between "system" libraries and user mappings and don't copy those system libraries to the mixxx user folder.

**Comment by daschuer** [comment]
As a band aid, can we detect this situation and fix at least the misleading error message?

**Comment by daschuer** [comment]
What is the situation here? We have postponed it form 2.5-alpa. At least the band aid looks like a low hanging fruit. 
Or has this issue gradually disappeared and this issue can be closed without. Can a controller expert have a look?

**Comment by acolombier** [comment]
Just making sure I understood the problem:
- One copies a mapping file (either common library or actually mapping script) in their `.mixxx` folder at version X
- They update to version X+1, which contains new changes in common library or mapping script, previously copied
- User still run the version from X and miss latest feature/bug fix

If I got this right, it sounds like there is two potential changes we could do:

1. Only use user defined library (e.g component JS) for user defined mapping. (easy)
2. Track recently changed controller mapping/lib in each version, and inform the user of first launch that they may be missing latest change due to their local copy, with a call to action to "archive" user define and make the use of the newly updated on instead (not sure if this is a clear, but a popup at first start following an update to tell them something like "Mapping ZZZ was updated in X+1, however you have a user defined on. Do you want to archive yours and use the system or keep using yours, at your own risk", medium-to-hard effort)

**Comment by daschuer** [comment]
What is the status here?

**Comment by ywwg** [comment]
suggest we punt to 2.7
